test/clj/game/cards/upgrades_test.clj

```diff
@@ -3336,3 +3336,243 @@
+      (click-prompt state :corp "3")                        ;;Boost trace by 3
+      (is (= 8 (:strength (get-prompt state :runner))) "3 base, +2 of upgrade +3 boost"))))
+
+(deftest akitaro-watanabe
+  (do-game
+    (new-game {:corp {:hand ["Akitaro Watanabe" "Ice Wall" "Enigma"]}})
+    (play-from-hand state :corp "Akitaro Watanabe" "HQ")
+    (play-from-hand state :corp "Ice Wall" "HQ")
+    (play-from-hand state :corp "Enigma" "HQ")              ;;-1c
+    (let [akitaro (get-content state :hq 0)
+          iw (get-ice state :hq 0)
+          enigma (get-ice state :hq 1)]
+      (rez state :corp akitaro)
+      (rez state :corp iw)
+      (rez state :corp enigma)
+      (is (= 2 (:credit (get-corp))) "2c left after installing and rezzing with AW bonus"))))
+
+(deftest research-station
+  (do-game
+    (new-game {:corp {:id "NBN: The World is Yours*"
+                      :hand [(qty "Research Station" 9)] :deck ["Research Station"]}})
+    (play-from-hand state :corp "Research Station" "HQ")
+    (rez state :corp (get-content state :hq 0))
+    (take-credits state :corp)
+    (is (= 8 (count (:hand (get-corp)))) "+3 cards, 2 of Research Station and +1 of ID")
+    (is (= nil (get-prompt state :corp)) "No prompt asking to discard cards from hand")
+    (take-credits state :runner)
+    (is (= 9 (count (:hand (get-corp)))) "Double check that you start next turn with 8 cards")))
+
+(deftest traffic-analyzer
+  (do-game
+    (new-game {:corp   {:hand ["Traffic Analyzer" "Ice Wall"]}
+               :runner {:tag 2 :hand [(qty "Corroder" 2)]}})
+    (play-from-hand state :corp "Traffic Analyzer" "HQ")
+    (play-from-hand state :corp "Ice Wall" "HQ")
+    (rez state :corp (get-content state :hq 0))
+    (rez state :corp (get-ice state :hq 0))
+    (click-prompt state :corp "4")
+    (click-prompt state :runner "0")
+    (is (= 1 (:credit (get-corp))) "After using all credits to boost trace, Corp gains 1c thanks to Traffic Analyzer")
+    ))
+
+(deftest defense-construct
+  (do-game
+    (new-game {:corp {:hand ["Defense Construct" (qty "PAD Campaign" 3)]}})
+    ;; Play 3 PAD Campaigns on the same server so 2 of them are trashed facedown
+    (play-from-hand state :corp "Defense Construct" "Archives")
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (click-prompt state :corp "OK")
+    (take-credits state :runner)
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (click-prompt state :corp "OK")
+    ;; Advance Defense Construct twice to recover both PAD Campaigns from discard
+    (let [dc (get-content state :archives 0)
+          advance-tokens 2]
+      (rez state :corp dc)
+      (advance state dc advance-tokens)
+      (take-credits state :corp)
+      (run-on state :archives)
+      (card-ability state :corp dc 0)
+      (click-card state :corp (get (:discard (get-corp)) 0))
+      (click-card state :corp (get (:discard (get-corp)) 1))
```

Comment:
```suggestion
      (click-card state :corp (first (:discard (get-corp))))
```
This works because `get-corp` will update each call, so the "top" card will be different.